src/func/distance.rs

```diff
@@ -0,0 +1,88 @@
+use serde::{Deserialize, Serialize};
+
+use super::{err::Error, Vector};
+
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
+pub enum Distance {
+    Dot,
+    Euclidean,
+    Cosine,
+}
+
+impl Distance {
+    pub fn from(distance: &String) -> Result<Self, Error> {
+        match distance.as_str() {
+            "dot" => Ok(Distance::Dot),
+            "euclidean" => Ok(Distance::Euclidean),
+            "cosine" => Ok(Distance::Cosine),
+            _ => Err("Distance not supported".into()),
+        }
+    }
+
+    pub fn calculate(&self, a: &Vector, b: &Vector) -> f32 {
+        assert_eq!(a.0.len(), b.0.len());
+
+        match self {
+            Distance::Dot => {
+                a.0.iter().zip(b.0.iter()).map(|(x, y)| x * y).sum::<f32>()
+            }
+            Distance::Euclidean => {
+                a.0.iter()
+                    .zip(b.0.iter())
+                    .map(|(a, b)| (a - b).powi(2))
+                    .sum::<f32>()
+                    .sqrt()
+            }
+            Distance::Cosine => {
+                let dot_product: f32 =
+                    a.0.iter().zip(b.0.iter()).map(|(x, y)| x * y).sum();
+                let magnitude_a: f32 =
+                    a.0.iter().map(|x| x.powi(2)).sum::<f32>().sqrt();
+                let magnitude_b: f32 =
+                    b.0.iter().map(|y| y.powi(2)).sum::<f32>().sqrt();
+
+                dot_product / (magnitude_a * magnitude_b)
```

Comment:
Can we have this in separate function like `calculate_cosine`?

Reply:
Sure